lib/rules/best-practises/reason-string.js

```diff
@@ -0,0 +1,101 @@
+const _ = require('lodash')
+const BaseChecker = require('./../base-checker')
+
+const DEFAULT_MAX_CHARACTERS_LONG = 20
+
+const ruleId = 'reason-string'
+const meta = {
+  type: 'best-practises',
+
+  docs: {
+    description:
+      'Require or revert statement must have a reason string and check that each reason string is at most N characters long.',
+    category: 'Best Practise Rules'
+  },
+
+  isDefault: false,
+  recommended: false,
+  defaultSetup: ['warn', { maxLength: DEFAULT_MAX_CHARACTERS_LONG }],
+
+  schema: [
+    {
+      type: 'array',
+      items: [
+        {
+          properties: {
+            maxLength: {
+              type: 'string'
+            }
+          },
+          additionalProperties: false
+        }
+      ],
+      uniqueItems: true,
+      minItems: 2
+    }
+  ]
+}
+
+class ReasonStringChecker extends BaseChecker {
+  constructor(reporter, config) {
+    super(reporter, ruleId, meta)
+
+    this.maxCharactersLong =
+      (config &&
+        config.getObjectPropertyNumber(ruleId, 'maxLength', DEFAULT_MAX_CHARACTERS_LONG)) ||
+      DEFAULT_MAX_CHARACTERS_LONG
+  }
+
+  enterFunctionCallArguments(ctx) {
+    if (this.isReasonStringStatement(ctx)) {
+      const functionParameters = this.getFunctionParameters(ctx)
+      const functionName = this.getFunctionName(ctx)
+
+      const lastFunctionParameter = _.last(functionParameters)
+
+      const haveReasonMessage = this.haveReasonMessage(lastFunctionParameter)
+      const haveNoMessage = functionParameters.length <= 1 || !haveReasonMessage
```

Comment:
The logic here is a little hard to follow. I think it can be improved by doing something like this:

```
if onlyOneParameter:
  raise noReasonMessage
  return
if lastParameterTooLong
  raise reasonTooLong
```

This won't check that the last parameter is a string, but that's probably a compiler's responsibility.

Reply:
Logic updated